### PR TITLE
Tracing panic hook

### LINKED ISSUE #3468: Log panics via `tracing`
On pageserver at least, assuming elsewhere as well, we rely on [`std::panicking::default_hook`](https://doc.rust-lang.org/stable/src/std/panicking.rs.html#235) panic hook for printing the panic to stderr. This is not great, because the stderr writes can get interleaved by the stdout writes, and we end up with botched log messages in grafana where there is a bit of the stacktrace in different messages.

What we should do instead is to install a panic hook *after* configuring logging *before* configuring sentry (which also installs a panic hook, which we don't want to override).

This should solve our interleaved messages in grafana issue, as well as "missing span at panic" issue.

This should be considered in other binaries as well, but perhaps easier to test with one.

Additionally it would be nice to have a panic endpoint on http which we could use to ensure both tracing and sentry work on staging and prod.

Relevant:
- [`tracing` has a panic hook example](https://github.com/tokio-rs/tracing/blob/ee8fd2876d5c3459ca4cce38ace29ea222a2a5f5/examples/examples/panic_hook.rs)
- [the std impl, but in release 1.66.1](https://github.com/rust-lang/rust/blob/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs#L235-L288)

### PR DESCRIPTION
**Pull request by koivunej**
Fixes #3468.

This does change how the panics look, and most importantly, make sure they are not interleaved with other messages. Adds a `GET /v1/panic` endpoint for panic testing (useful for sentry dedup and this hook testing).

The panics are now logged within a new error level span called `panic` which separates it from other error level events. The panic info is unpacked similar to in tracing example with fields:
- thread=mgmt request worker
- location="pageserver/src/http/routes.rs:898:9"
  - instead of:
      - panic.file="pageserver/src/http/routes.rs"
      - panic.line=898
      - panic.column=9

Much interest was not sparked, I am thinking of deciding the below as:
- no frame filtering until std supports it (same for head, tail)
- no getting rid of trailing `\n` (splitting should not care -- lets try this on staging)
- location formatting changed (both are above)
- try first on pageserver on staging because we can test this (added http route)

Decision points:
- [x] should we filter the first frames up to `rust_begin_unwind`?
- [x] should we filter the last frames up with same heuristics as anyhow? (I haven't checked what those are)
- [x] should we get rid of the trailing `\n` on the message? (re: message splitting for promtail)
- [x] should we change the location formatting?
- [x] move directly to utils under logging?
- [x] apply to all binaries directly or first test with `pageserver`?

### CODE REVIEW

**Comment by koivunej** [comment]
Old panic looked like (via `GET /v1/panic`) with prev and next line:

```
2023-01-27T16:28:14.440275Z  INFO GET /v1/always_panic 404 Not Found
thread 'mgmt request worker' panicked at 'unconditional panic for testing panic hook integration', pageserver/src/http/routes.rs:898:9
stack backtrace:
   0: rust_begin_unwind
--snip--
  64: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at /home/joonas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.2/src/runtime/blocking/pool.rs:469:13
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
2023-01-27T16:28:26.059654Z  INFO GET /v1/panic 204 No Content
```

New panic looks like (with anyhow backtrace above):

```
             at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
2023-01-27T16:56:48.866015Z ERROR panic{thread=mgmt request worker panic.file="pageserver/src/http/routes.rs" panic.line=898 panic.column=9}: unconditional panic for testing panic hook integration

Stack backtrace:
   0: pageserver::tracing_panic_hook
             at /home/joonas/src/neon/neon-main/pageserver/src/bin/pageserver.rs:503:21
   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/alloc/src/boxed.rs:2001:9
   2: std::panicking::rust_panic_with_hook
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:692:13
   3: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:577:13
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/sys_common/backtrace.rs:137:18
   5: rust_begin_unwind
--snip--
  82: clone3
             at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81

2023-01-27T16:56:48.970587Z  INFO GET /v1/panic 204 No Content
```

Above also shows the new frames.

So there's an extra newline *after*. I wonder if that's a problem for message splitting, we could try to make it go away by depending directly on `backtrace-rs` instead of `std` (like `anyhow`) but I would prefer not doing that.

Could @lassizci comment on if this will create an issue for the message splitting pipeline, or point me to the implementation?

**Comment by lassizci** [comment]
> Could @lassizci comment on if this will create an issue for the message splitting pipeline, or point me to the implementation?

@koivunej Shouldn't affect, the matcher seems to be matching a timestamp and then collecting the lines until the next timestamp.

**Comment by koivunej** [comment]
Postponing this.

File: pageserver/src/bin/pageserver.rs [hunk 1]
```diff
@@ -495,6 +499,50 @@ fn cli() -> Command {
...
+        Some(s) => *s,
+        None => match info.payload().downcast_ref::<String>() {
+            Some(s) => &s[..],
+            None => "Box<dyn Any>",
+        },
+    };
+
+    let thread = std::thread::current();
+    let thread = thread.name().unwrap_or("<unnamed>");
+    let backtrace = std::backtrace::Backtrace::capture();
```
Comment (problame) [inline pageserver/src/bin/pageserver.rs:518]:
```suggestion
    # Use force_capture: we want the stack trace in the logs, regardless of
    # whether RUST_BACKTRACE is set or not.
    let backtrace = std::backtrace::Backtrace::force_capture();
```

File: pageserver/src/http/routes.rs [hunk 2]
```diff
@@ -1029,6 +1029,16 @@ async fn active_timeline_of_active_tenant(
         .map_err(ApiError::NotFound)
 }
 
+async fn always_panic_handler(_: Request<Body>) -> Result<Response<Body>, ApiError> {
+    // catch_unwind ensures that tokio nor hyper are distracted by our panic, but panic hooks
+    // (at least sentry) will report the issue. if tested across fleet, this should be a good test
+    // case for panic deduplication as well per version.
```
Comment (problame) [inline pageserver/src/http/routes.rs:1035]:
```suggestion
    // Deliberately cause a panic to exercise the panic hook registered via std::panic::set_hook().
    // For pageserver, the relevant panic hook is `tracing_panic_hook` , and the `sentry` crate's wrapper around it.
    // Use catch_unwind to ensure that tokio nor hyper are distracted by our panic.
```

File: pageserver/src/http/routes.rs [hunk 3]
```diff
@@ -1029,6 +1029,16 @@ async fn active_timeline_of_active_tenant(
...
+    let _ = std::panic::catch_unwind(|| {
+        panic!("unconditional panic for testing panic hook integration")
```
Comment (problame) [inline pageserver/src/http/routes.rs:1037]: How about repeating a caller-provided string in the panic message?
Caller can still use a constant, but that way, they can pass in a nonce and check the logs / sentry for that exact nonce.

**Review by problame — APPROVED** [review]
Generally LGTM, minor nits and one improvement suggestion.

File: pageserver/src/bin/pageserver.rs (hunk 1 above)
Comment (koivunej) [reply pageserver/src/bin/pageserver.rs:518]: We have that setting already set in deployments, what if you are manual testing and want to hit something often? That'd be altering the existing behaviour, so don't like it.

File: pageserver/src/http/routes.rs (hunk 3 above)
Comment (koivunej) [reply pageserver/src/http/routes.rs:1037]: I'll add an optional message query parameter.

File: pageserver/src/http/routes.rs (hunk 2 above)
Comment (koivunej) [reply pageserver/src/http/routes.rs:1035]: Is this just stylistic rewrite or is there some changes I cannot see? unsure if std's panic hook needs to be mentioned, it's the only panic hook in std..

File: pageserver/src/http/routes.rs (hunk 3 above)
Comment (koivunej) [reply pageserver/src/http/routes.rs:1037]: I'll just dump the whole url decoded query actually.

File: pageserver/src/http/routes.rs (hunk 2 above)
Comment (koivunej) [reply pageserver/src/http/routes.rs:1035]: Hmm ok, now I can see.

File: pageserver/src/http/routes.rs (hunk 3 above)
Comment (koivunej) [reply pageserver/src/http/routes.rs]: [9d8e796](https://github.com/neondatabase/neon/pull/3475/commits/9d8e796e866ed4e6f5b9f2f4ab98f00b6ec0c47d)

File: pageserver/src/bin/pageserver.rs (hunk 1 above)
Comment (koivunej) [reply pageserver/src/bin/pageserver.rs:518]: Resolving.

**Pull request merged**